### ISSUE TITLE
"Reply channel not set up yet" when 2 HTTP Citrus actions happen in parallel

### ISSUE THREAD

**Issue by progaddict** [citrusframework/citrus#550]
**Citrus Version**
2.7.8

**Expected behavior**
Test passes

**Actual behavior**
Test fails

**Test case description**
2 logically identical async blocks with different test data, pseudo code:
```
async().actions(
  insert data 1 into DB,
  behaviour with HTTP receive() and respond() with HTTP 200 OK,
  DB check that a record with status "SUCCESS" exists for the test data 1
);

async().actions(
  insert data 2 into DB,
  behaviour with HTTP receive() and respond() with HTTP 200 OK,
  DB check that a record with status "SUCCESS" exists for the test data 2
);
```

SUT behavior:
* SUT has a background job which is triggered every 5 sec (on TEST env)
* The background job scans DB for new entries and starts processing them
* SUT then does HTTP POST with this new data as payload to an external system (Citrus HTTP mock server)
* The external system responds with some additional data (the mock replies with HTTP 200 OK)
* When SUT receives this data, the processing is finished and SUT then stores a record with status SUCCESS into DB (Citrus checks that DB has this entry at the end of the `async` blocks)

Note: one `async` block alone works but 2 blocks cause a failure:
```
12:57:06,068 DEBUG on.DefaultCorrelationManager| Finding correlated object for 'citrus_message_id = '9a369286-4824-4d6c-9eac-0bb12308d19a''
12:57:06,068 DEBUG           citrus.RetryLogger| Reply channel not set up yet - retrying in 500ms
```
It seems that Citrus cannot find the right HTTP channel for sending a response.

**Comment by CCarmenR** [comment]
I think the issue is neither related to the message "Reply channel not set up yet" nor to two HTTP Citrus actions happening in parallel. The tests that end with "CitrusRuntimeException: Failed to wait for test container to finish properly" are also sometimes green (in about 3 of 10 times). Log excerpt when tests are green: 
```
11:27:45,509 DEBUG citrus.RetryLogger| No message received with message selector - retrying in 500ms
11:27:45,520 DEBUG on.DefaultCorrelationManager| Finding correlated object for 'citrus_message_id = 'ec764b1e-a318-4913-9d03-06f60474377c''
11:27:45,520 DEBUG citrus.RetryLogger| Reply channel not set up yet - retrying in 500ms
11:27:45,590 DEBUG or.LoggingHandlerInterceptor| Received Http request:
HTTP/1.1 POST /casa/api/contracts/v2/search 
```
I could reproduce the blinking test behavior resulting in the same message in a test case which first sets up a Citrus behavior receiving and responding to an HTTP request and then expects to receive a JMS message - the latter in an async block.  Moving the async block above applyBehavior for the HTTP server setup the problem did not occur anymore.

**Comment by progaddict** [comment]
Hi @svettwer  !

I had some spare time to implement a super-duper-small example SUT and a couple of Citrus tests which reproduce this bug: https://github.com/progaddict/citrus-multiple-async-blocks

I hope this will help you guys to fixie-wixie the bug super-duper-fast :laughing:

**Comment by christophd** [comment]
You need to set up a manual message correlation in that particular use case. Citrus is able to auto correlate request and response on one endpoint in a test case. In case you have parallel message exchange on the same endpoint in one test case you need manual correlation.

You can see this in:
https://github.com/citrusframework/citrus/blob/master/modules/citrus-integration/src/it/resources/com/consol/citrus/http/HttpParallelRequest_3_IT.xml

**Comment by christophd** [comment]
Your test set up looks pretty weird to me though. Why not removing both async blocks and inserting both DB entries after each other. then receive incoming messages with a message selector and at the end verify that DB entries were updated too.

**Comment by svettwer** [comment]
Hi!

Yeah, one could discuss whether the test case is meaningful but from a technical perspective, this scenario should be testable. Nevertheless, I have a feeling that this will be a tricky one to analyze. 

BR,
Sven

**Comment by svettwer** [comment]
Thx to @progaddict and his SUT project, I was able to shrink the context of the issue a little bit further.

**TL;DR**
It seems that multiple `async` container disturb each other in some way. It might be a race condition or a message overwriting in the message channel depending on the order the messages arrive.

**Details**
I've created the following test cases:
```java
public class TodoListIT extends TestNGCitrusTestRunner {

    @Autowired
    private HttpClient httpClient;

    @Autowired
    private HttpServer httpServer;

    @Test
    @CitrusTest
    public void testFailing() {

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/foo")
                        .payload(createPayload("foo"))),

                http(b -> b.client(httpClient).
                        receive()
                        .response(HttpStatus.OK)),

                simulateBackend("foo")

        ).addErrorAction(echo("foo failed"));

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/boo")
                        .payload(createPayload("boo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)),

                simulateBackend("boo")
        ).addErrorAction(echo("boo failed"));

    }

    @Test
    @CitrusTest
    public void testPassingAsyncParallelSequential() {

        parallel().actions(
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/foo")
                                .payload(createPayload("foo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)),

                        simulateBackend("foo")
                )
        );

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/boo")
                        .payload(createPayload("boo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)),

                simulateBackend("boo")
        );

    }

    @Test
    @CitrusTest
    public void testPassingParallelSequentialParallelSequential() {

        parallel().actions(
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/foo")
                                .payload(createPayload("foo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)),
                        simulateBackend("foo")
                )
        );

        parallel().actions(
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/boo")
                                .payload(createPayload("boo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)),
                        simulateBackend("boo")
                )
        );
    }

    private String createPayload(final String boo) {
        return String.format("{\"name\": \"%s\" }", boo);
    }

    private ApplyTestBehaviorAction simulateBackend(final String payload) {
        return applyBehavior(new AbstractTestBehavior(){

            @Override
            public void apply() {
                http(b -> b.server(httpServer)
                        .receive()
                        .post("/" + payload)
                        .selector(Collections.singletonMap("jsonPath:$.name", payload)));

                http(b -> b.server(httpServer)
                        .respond()
                        .status(HttpStatus.OK));
            }
        });
    }
}
```

with the config:
```java
@Configuration
public class EndpointConfig {

    @Bean
    public HttpClient todoClient() {
        return CitrusEndpoints.http()
                .client()
                .requestUrl("http://localhost:8888")
                .build();
    }

    @Bean
    public HttpServer todoServer() {
        return CitrusEndpoints.http()
                .server()
                .port(8888)
                .autoStart(true)
                .build();
    }
}
```

The sample project configured the endpoints differently, with a higher timeout than the `async` container comes with. That's why the error message is misleading and the endpoint is polling much longer as the container runs. This is especially important for #548.

I've reduced the high timeouts which leads to a different, more test action specific error message.

At the end, it seems like the multiple `async` container disturb each other. If you execute the failing test multiple times, you'll notice that the issue occurs sometimes for `foo` and sometimes for `boo`. This could indicate a race condition or a message overwriting in the message channel depending on the order the messages arrive.

I'll have to investigate this a little bit further.

**Comment by svettwer** [comment]
This may also be related to #542.

**Comment by svettwer** [comment]
Btw: The [async documentation](http://citrusframework.org/citrus/reference/html/index.html#containers-async) states that the on error and on success callbacks are just available in XML DSL, which is not the case anymore.

**Comment by christophd** [comment]
@svettwer the two passing sample tests that you have provided do not use parallel http requests. 

In `testPassingParallelSequentialParallelSequential()` you should not use two parallel blocks. Instead use just one single parallel with all http requests in it in order to have a real parallel behavior as the test is supposed to do.

Also you should not separate async and parallel in `testPassingAsyncParallelSequential()` in order to have parallel http requests in your test.

At the moment the two passing tests are using sequential http requests and therefore pass the test. When you adjust the tests as suggested above they will fail, too.

**Comment by christophd** [comment]
So async and parallel are both not able to handle multiple http requests out of the box. This is a known limitation that needs to be fixed with manual correlation as described in https://github.com/citrusframework/citrus/blob/master/modules/citrus-integration/src/it/resources/com/consol/citrus/http/HttpParallelRequest_3_IT.xml

**Comment by christophd** [comment]
I think documentation already points to this limitation isn't it?

**Comment by christophd** [comment]
Here is my adjusted sample that is failing consistently with async and parallel:

```java
public class TodoListIT extends TestNGCitrusTestRunner {

    @Autowired
    private HttpClient httpClient;

    @Autowired
    private HttpServer httpServer;

    @Test
    @CitrusTest
    public void testAsync() {

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/foo")
                        .payload(createPayload("foo"))),

                http(b -> b.client(httpClient).
                        receive()
                        .response(HttpStatus.OK)),

                simulateBackend("foo")

        ).addErrorAction(echo("foo failed"));

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/boo")
                        .payload(createPayload("boo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)),

                simulateBackend("boo")
        ).addErrorAction(echo("boo failed"));

    }

    @Test
    @CitrusTest
    public void testAsyncParallelSequential() {

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/foo")
                        .payload(createPayload("foo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)),

                simulateBackend("foo")
        );

        http(b -> b.client(httpClient)
                .send()
                .post("/boo")
                .payload(createPayload("boo")));

        http(b -> b.client(httpClient)
                .receive()
                .response(HttpStatus.OK));

        simulateBackend("boo");
    }

    @Test
    @CitrusTest
    public void testParallelSequentialParallelSequential() {

        parallel().actions(
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/foo")
                                .payload(createPayload("foo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)),
                        simulateBackend("foo")
                ),
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/boo")
                                .payload(createPayload("boo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)),
                        simulateBackend("boo")
                )
        );
    }

    private String createPayload(final String boo) {
        return String.format("{\"name\": \"%s\" }", boo);
    }

    private ApplyTestBehaviorAction simulateBackend(final String payload) {
        return applyBehavior(new AbstractTestBehavior(){

            @Override
            public void apply() {
                http(b -> b.server(httpServer)
                        .receive()
                        .post("/" + payload)
                        .selector(Collections.singletonMap("jsonPath:$.name", payload)));

                http(b -> b.server(httpServer)
                        .respond()
                        .status(HttpStatus.OK));
            }
        });
    }
}
```

**Comment by christophd** [comment]
I finally came up with following sample that demonstrates the manual correlation:

Use Http server with static response
--- 
```java
@Bean
    public HttpServer todoServer() {
        return CitrusEndpoints.http()
                .server()
                .port(8888)
                .autoStart(true)
                .endpointAdapter(new StaticEndpointAdapter() {
                    @Override
                    protected Message handleMessageInternal(Message message) {
                        return new HttpMessage(message.getPayload())
                                .status(HttpStatus.OK);
                    }
                })
                .build();
    }
```

Tests with manual correlation
---
```java
public class TodoListIT extends TestNGCitrusTestRunner {

    @Autowired
    private HttpClient httpClient;

    @Test
    @CitrusTest
    public void testAsync() {

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/foo")
                        .extractFromHeader(MessageHeaders.ID, "request#1")
                        .payload(createPayload("foo"))),

                http(b -> b.client(httpClient).
                        receive()
                        .response(HttpStatus.OK)
                        .payload(createPayload("foo"))
                        .selector(Collections.singletonMap(MessageHeaders.ID, "${request#1}")))

        ).addErrorAction(echo("foo failed"));

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/boo")
                        .extractFromHeader(MessageHeaders.ID, "request#2")
                        .payload(createPayload("boo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)
                        .payload(createPayload("boo"))
                        .selector(Collections.singletonMap(MessageHeaders.ID, "${request#2}")))
        ).addErrorAction(echo("boo failed"));

    }

    @Test
    @CitrusTest
    public void testAsyncParallelSequential() {

        async().actions(
                http(b -> b.client(httpClient)
                        .send()
                        .post("/foo")
                        .extractFromHeader(MessageHeaders.ID, "request#1")
                        .payload(createPayload("foo"))),

                http(b -> b.client(httpClient)
                        .receive()
                        .response(HttpStatus.OK)
                        .payload(createPayload("foo"))
                        .selector(Collections.singletonMap(MessageHeaders.ID, "${request#1}")))
        );

        http(b -> b.client(httpClient)
                .send()
                .post("/boo")
                .extractFromHeader(MessageHeaders.ID, "request#2")
                .payload(createPayload("boo")));

        http(b -> b.client(httpClient)
                .receive()
                .response(HttpStatus.OK)
                .payload(createPayload("boo"))
                .selector(Collections.singletonMap(MessageHeaders.ID, "${request#2}")));
    }

    @Test
    @CitrusTest
    public void testParallelSequentialParallelSequential() {

        parallel().actions(
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/foo")
                                .extractFromHeader(MessageHeaders.ID, "request#1")
                                .payload(createPayload("foo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)
                                .payload(createPayload("foo"))
                                .selector(Collections.singletonMap(MessageHeaders.ID, "${request#1}")))
                ),
                sequential().actions(
                        http(b -> b.client(httpClient)
                                .send()
                                .post("/boo")
                                .extractFromHeader(MessageHeaders.ID, "request#2")
                                .payload(createPayload("boo"))),

                        http(b -> b.client(httpClient)
                                .receive()
                                .response(HttpStatus.OK)
                                .payload(createPayload("boo"))
                                .selector(Collections.singletonMap(MessageHeaders.ID, "${request#2}")))
                )
        );
    }

    private String createPayload(final String boo) {
        return String.format("{\"name\": \"%s\" }", boo);
    }
}
```

**Comment by christophd** [comment]
All tests are green with this correlation in place

**Comment by svettwer** [comment]
Hi!

>So async and parallel are both not able to handle multiple http requests out of the box. This is a known limitation that needs to be fixed with manual correlation as described in [...]

The manual correlation might fix it but I still don't get why this is required. So the header `citrus_message_id` holds a correlation id that identifies the message internally within the communication. Why do I have to map this manually? This could be done complete behind the scenes, right? I mean the only use case for such a loose coupling I can imagine is, when I want to receive a response with a different correlation id as sent in the request.

>I think documentation already points to this limitation isn't it?

If it is, four people were unable to find it. :sweat_smile:
Could you provide a link @christophd? 

> another problem in your sample tests is the mix of client and server actions in the very same test without accounting for synchronous nature of http as messaging protocol.

Well, I do know about the *synchronous nature of http as messaging protocol* but yes, the sample was not as clean as it should have been. Mea culpa.

>Your client requests will not get an immediate proper response from server but force a timeout on the server and get a default Http OK 200 then as response after timeout.

`200` is the default response of a timeout? Wouldn't `408` be more appropriate. I think, this could lead to some false positives as it did here. :thinking: 

But btt:
Even if Citrus is designed that way, I think it would be more intuitive for the user if the correlation ID mapping would happen behind the scenes so that parallel requests work out of the box. Because I as a framework user, would not want to care about manual mapping of internal IDs of the framework.
So from my perspective I'd be okay that this is a enhancement and not a bug.

BR,
Sven

**Comment by svettwer** [comment]
Hi!

>The manual correlation might fix it but I still don't get why this is required. So the header citrus_message_id holds a correlation id that identifies the message internally within the communication. Why do I have to map this manually? This could be done complete behind the scenes, right?

Okay, I see that we have to close the gap between the send and receive messages concerning the correlation ID...

**Comment by christophd** [comment]
> Could you provide a link @christophd?
Definitely a documentation issue ;-)

**Comment by christophd** [comment]
> 200 is the default response of a timeout? Wouldn't 408 be more appropriate.

Yes, for a productive server this would definitely apply. Here we have a mocking server with some default behavior that happens to be the default positive behavior on purpose so users may skip the response definition in tests.

As 200 OK is the mainly used response code so I think the default behavior of the server mock should apply to that.

**Comment by christophd** [comment]
> Why do I have to map this manually? This could be done complete behind the scenes, right? I mean the only use case for such a loose coupling I can imagine is, when I want to receive a response with a different correlation id as sent in the request.

I think doing this behind the scenes is not an easy task, yet possible with some FIFO assumption on multiple requests. I think manual correlation should remain as fallback, too as in some scenarios testers need to control the message flow in a very specific order. For instance the 2nd request might be handled in validation before the 1st request on purpose.

**Comment by svettwer** [comment]
closed by #591
